src/shared/services/HttpService.ts

```diff
@@ -0,0 +1,92 @@
+import { LemmyHttp } from "lemmy-js-client";
+import { getHttpBase } from "../../shared/env";
+import { i18n } from "../../shared/i18next";
+import { toast } from "../../shared/utils";
+
+type EmptyRequestState = {
+  state: "empty";
+};
+
+type LoadingRequestState = {
+  state: "loading";
+};
+
+type FailedRequestState = {
+  state: "failed";
+  msg: string;
+};
+
+type SuccessRequestState<T> = {
+  state: "success";
+  data: T;
+};
+
+/**
+ * Shows the state of an API request.
+ *
+ * Can be empty, loading, failed, or success
+ */
+export type RequestState<T> =
+  | EmptyRequestState
+  | LoadingRequestState
+  | FailedRequestState
+  | SuccessRequestState<T>;
+
+export async function apiWrapper<ResponseType>(
+  req: Promise<ResponseType>
+): Promise<RequestState<ResponseType>> {
+  try {
+    const res = await req;
+    return {
+      state: "success",
+      data: res,
+    };
+  } catch (error) {
+    console.error(`API error: ${error}`);
+    toast(i18n.t(error), "danger");
+    return {
+      state: "failed",
+      msg: error,
+    };
+  }
+}
+
+/**
+ * A Special type of apiWrapper, used only for the iso routes.
+ *
+ * Necessary because constructors can't be async
+ */
+export function apiWrapperIso<ResponseType>(
+  res: ResponseType
+): RequestState<ResponseType> {
+  try {
+    return {
```

Comment:
@dessalines I pointed this out on one of the draft PRs. I have no idea what error could be thrown in this try block.

Reply:
lemmy-js-client throws an API error, returns whatever is in the `result[error]` as a string, and this catches it, and returns a failed state with that message.

Reply:
I understand, however the API call isn't being done here. The function takes an already returned response type as an argument.

Reply:
Ah I gotcha... yep the isoWrapper version since its not async (and can't be), can really only take a sucessful response.

Reply:
This comment change was resolved by centralizing the API wrapping logic in `HttpService.ts`.